reactive/ovn_chassis_charm_handlers.py

```diff
@@ -73,6 +73,19 @@ def enable_openstack():
         charm_instance.assess_status()
 
 
+@reactive.when(OVN_CHASSIS_ENABLE_HANDLERS_FLAG)
+@reactive.when_none('charm.installed')
+def maybe_disable_mlockall():
+    with charm.provide_charm_instance() as charm_instance:
+        if charm_instance.options.mlockall_disabled:
+            # We need to render /etc/default/openvswitch-switch after the
+            # initial install and restart openvswitch-switch. This is done to
+            # ensure that when the disable-mlockall config option is unset,
+            # mlockall is disabled when running in a container. LP: #1906280
+            charm_instance.render_openvswitch_defaults(
+                restart=not reactive.is_flag_set('charm.paused'))
```

Comment:
I don't think openvswitch-switch is stopped during the pause action. Does that affect this logic? I'm thinking in that case that once resumed, the charm still won't pick up the mlockall setting.

Reply:
Good question, the charm will do a full hook execution on resume:
https://github.com/openstack-charmers/charm-layer-ovn/blob/b5a86ead9e130880a8a9a8c6c5cc861aeef3e9eb/lib/charms/ovn_charm.py#L305-L323

That in turn will call configure_ovs() which will do the restart if the disable-mlockall config is changed.

I guess the question then becomes:
- Does the config.changed flag stay set through the charm not responding to it while paused?
- If the charm is deployed to a already working unit with neutron-openvswitch on it, would that not mean it either does not need the mlockall handling or neutron-openvswitch has already dealt with it?

Reply:
Good questions. One of the issues is that disable-mlockall defaults are different based on whether it's running in a container or not. In other words if disable-mlockall is unset and the charm is running in a container, disable-mlockall defaults to true behavior (and false if not in a container). tl;dr there are default cases where we need to disable mlockall and restart ovs when config.changed == false. neutron-ovs has the same behavior so post-ugprade should have the same mlockall behavior.

Reply:
Light bulb, so yes this might have been handled already by neutron-ovs and not require a restart on upgrade from ovs to ovn. Although I don't know how we can guarantee that it is an upgrade from ovs->ovn just because the charm was deployed in a paused state. hmm.

Reply:
OK, so as a compromise I propose we do the on-install restart when either being in a container or `disable-mlockall` config being explicitly set, without taking paused state into account. Otherwise this becomes way too complicated.

Reply:
That makes sense. I hate to restart ovs if it doesn't absolutely need to, but I think in this case we need to make a bit of an assumption.